design-documents/service-isolation/authentication-authorization.md

```diff
@@ -0,0 +1,80 @@
+     - Pros: services control all the flow of roles and privileges
+     - Cons: the same logic is duplicated in multiple services
+3. BFF/Monolith performs authentication, a service performs authorization
+    - Pros: single entry point for authentication, services control roles & privileges
+    - Cons: shared roles & privileges data between services
+4. BFF/Monolith performs authentication and authorization, a service uses a "secure" token
+    - Pros: a single entry point for credentials and permissions validation, more secure communication, avoids
+[Confused Deputy Problem](https://en.wikipedia.org/wiki/Confused_deputy_problem)
+    - Cons: additional logic to validate a "secure" token (could be extracted to a separate service)
+
+According to described above strategies, the last approach looks like more convenient. It allows building authentication
+& authorization flow with a single source of roles & privileges and provides additional secure communication for separate
+services.
+
+### Token-based Auth
+
+Token-based auth allows users to use credentials like username and password to obtain a token which allows communicating
+with the system without using credentials.
+As Monolith/BFF performs authentication & authorization token-based auth allows using tokens for a secure communication
+between a user, Monolith/BFF, and services.
+
+[JWT](https://jwt.io/) could be used as a "secure" token implementation and the main benefits are:
+* Allows using adaptive structure to pass permissions between services
+* Could be encrypted (JWE) or just signed (JWS)
+* As it's RFC it has implementations for most of the modern programming languages, so there is no a hard
+dependency on some stack for the service implementation
+    
+## General Schema
+
+As Magento should provide [multiple BFF](https://github.com/magento/architecture/blob/master/design-documents/service-isolation.md#backends-for-frontends)
+it makes sense to extract authentication and authorization flow to a separate `Auth Service` to avoid credentials
+validation and duplication of roles & privileges data. And the general schema might look like this:
+
+![Authentication and Authorization General Schema](auth-general-schema.png)
+
+And the basic scenario might be as following:
+
+![Auth Basic Scenario](auth-basic-scenario.png)
+
+Also, there are a few variants of the proposed schema like to have separate services for token validation and renewal:
+
+![Token Validation Service](token-validation-service.png)
+
+Now, each service validates a token through `Token Service` it reduces logic duplication but adds additional requests
+which might impact on performance.
+
+According to the proposed schema, user-Monolith/BFF scenario could use JWE tokens, Monolith/BFF - service could use JWS
+tokens to make the communication more secure.
+
+### OAuth 2.0
+
+As `Auth Service` is the single entry for authentication & authorization it should store roles and privileges for all
```

Comment:
I propose not to validate privileges when we have service to service communication. All operations will be defined on the BFF/monolith and these are operations we need to control access to. To perform these operations monolith may talk to one or more services which in turn can talk to other services. For these communication I propose not to check privileges. It's more work and we will be granting privileges to operations defined in BFF/monolith. We can do authentication if we want to make communication within the network more secure. Some people don't do it.

Reply:
How we can guarantee that service could execute a request from another service?

Reply:
It would be good to define cases when do we need to check if service can call other service to perform operation. How it's different from monolith deployment?

Some people don't perform this validation. Communication between services is trusted and secured from public by private network.

Reply:
> Some people don't perform this validation. Communication between services is trusted and secured from public by private network.

That was our first discussion, probably we need to discuss such approach again about private/public services communication.